pkg/kubelet/server/server.go

```diff
@@ -729,9 +729,40 @@ func (s *Server) getContainerLogs(request *restful.Request, response *restful.Re
 		response.WriteError(http.StatusInternalServerError, fmt.Errorf("unable to convert %v into http.Flusher, cannot show logs", reflect.TypeOf(response)))
 		return
 	}
-	fw := flushwriter.Wrap(response.ResponseWriter)
+
+	var (
+		stdout io.Writer
+		stderr io.Writer
+		fw     = flushwriter.Wrap(response.ResponseWriter)
+	)
+	if utilfeature.DefaultFeatureGate.Enabled(features.SplitStdoutAndStderr) {
+		wantedStream := logOptions.Stream
+		// No stream type specified, default to All
+		if wantedStream == nil {
+			allStream := v1.LogStreamTypeAll
+			wantedStream = &allStream
+		}
+		switch *wantedStream {
+		case v1.LogStreamTypeStdout:
+			stdout, stderr = fw, nil
+		case v1.LogStreamTypeStderr:
+			stdout, stderr = nil, fw
+		case v1.LogStreamTypeAll:
+			stdout, stderr = fw, fw
+		default:
+			_ = response.WriteError(http.StatusBadRequest, fmt.Errorf("invalid stream type %q", *logOptions.Stream))
+			return
+		}
+	} else {
+		if logOptions.Stream != nil && *logOptions.Stream != v1.LogStreamTypeAll {
+			_ = response.WriteError(http.StatusBadRequest, fmt.Errorf("unable to return the given log stream: %q. Please enable SplitStdoutAndStderr feature gate in kubelet", *logOptions.Stream))
+			return
+		}
+		stdout, stderr = fw, fw
+	}
+
 	response.Header().Set("Transfer-Encoding", "chunked")
-	if err := s.host.GetKubeletContainerLogs(ctx, kubecontainer.GetPodFullName(pod), containerName, logOptions, fw, fw); err != nil {
+	if err := s.host.GetKubeletContainerLogs(ctx, kubecontainer.GetPodFullName(pod), containerName, logOptions, stdout, stderr); err != nil {
```

Comment:
In `GetKubeletContainerLogs()`, it seems necessary to handle a case where `stdout` is `nil`:
https://github.com/kubernetes/kubernetes/blob/de8f6b0db7a6d7506b7544021be413d3faa47078/pkg/kubelet/kubelet_pods.go#L1555

Reply:
Nice catch. Thank you!